### PR TITLE
Support non-interactive user creation

### PR DESCRIPTION
**Pull request by nak3**
This supports --password option for user creation of CLI and stops prompting for the password.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3603)

### CODE REVIEW

**Comment by tschottdorf** [comment]
I wonder whether it'd feel less obviously unsafe to read the password from `stdin` instead? At least that way it wouldn't show up in `ps`, etc.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[cli/flags.go, line 231 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-K7GHOnhEiGDTA6XEJk0:-K7GHOnhEiGDTA6XEJk1:1514004646) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0aaf95e80bf3dccd11ef7c3d111423232c58d8d/cli/flags.go#L231)):*
why not replace this whole block with

```go
setUserCmd.Flags().StringVar(&password, "password", "", flagUsage["password"])
```

---

*[cli/flags.go, line 259 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-K7GHhVpUZtV8bhq-Z6N:-K7GHhVpUZtV8bhq-Z6O:1197679173) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0aaf95e80bf3dccd11ef7c3d111423232c58d8d/cli/flags.go#L259)):*
remove this newline.

---

*[cli/user.go, line 138 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-K7GI3TpTP170zIc3ywd:-K7GI3TpTP170zIc3ywe:394803613) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0aaf95e80bf3dccd11ef7c3d111423232c58d8d/cli/user.go#L138)):*
remove both the `log.Error` and `return` and instead put `panic(err)` (see https://github.com/cockroachdb/cockroach/pull/3581). Same a few lines up.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-:-K7GIHgpgp3SFz10R9U_:1671475744)*

**Comment by nak3** [comment]
I agree. I added the function to read from stdin with `password=-`.

---

Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[cli/flags.go, line 231 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-K7GHOnhEiGDTA6XEJk0:-K7HKvGLTf9JM1Q8MFlQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0aaf95e80bf3dccd11ef7c3d111423232c58d8d/cli/flags.go#L231)):*
Done.

---

*[cli/flags.go, line 259 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-K7GHhVpUZtV8bhq-Z6N:-K7HKz2qSU4Aq5sQIGow:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0aaf95e80bf3dccd11ef7c3d111423232c58d8d/cli/flags.go#L259)):*
Done.

---

*[cli/user.go, line 138 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-K7GI3TpTP170zIc3ywd:-K7HL-t_qa7-fIdMUexM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0aaf95e80bf3dccd11ef7c3d111423232c58d8d/cli/user.go#L138)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-:-K7HL1sVRv8bHAsfBomL:1665266952)*

**Comment by mberhault** [comment]
This seems fine but please be aware that the users table is not currently used by anything, it's ultimately meant for the web UI, and perhaps password-based authentication later. For now, all API calls are authenticated through SSL.

**Comment by tschottdorf** [comment]
Looks good, thanks! Can you add some examples/tests  in `cli/cli_test.go`? After that, good to go as far as I'm concerned.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-:-K7HNUm7BhmcmG1L-Cm-:-1671977807)*

**Comment by nak3** [comment]
@mberhault Yeah, that's true. Thank you.
@tschottdorf Thank you. I added test codes to `cli/cli_test.go`.

**Comment by tamird** [comment]
@nake3 we had a nasty bug on master that made CI not run our tests. Could you rebase this on latest master?

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603)*

**Comment by nak3** [comment]
@tamird It looks fine now. Or should I rebase it?

**Comment by tamird** [comment]
Please rebase

**Comment by nak3** [comment]
@tamird Sure, done.

File: cli/flags.go [hunk 1]
```diff
@@ -143,6 +143,9 @@ var flagUsage = map[string]string{
 		Determines the criteria used by nodes to make balanced allocation
 		decisions.  Valid options are "usage" (default) or "rangecount".
 `,
+	"password": `
+        Set password. It will disable prompting for the password. It reads from stdin if password=-
```
Comment (tamird) [inline cli/flags.go]: The other descriptions here use the present tense; how about:
```
The created user's password. If provided, disables prompting. Pass `-` to provide the password on standard input.
```

File: cli/user.go [hunk 2]
```diff
@@ -45,8 +47,7 @@ func runGetUser(cmd *cobra.Command, args []string) {
 	defer func() { _ = db.Close() }()
 	err := runPrettyQuery(db, os.Stdout, `SELECT * FROM system.users WHERE username=$1`, args[0])
 	if err != nil {
-		log.Error(err)
-		return
+		panic(err)
```
Comment (tamird) [inline cli/user.go]: @tschottdorf all these intended?

File: cli/user.go [hunk 3]
```diff
@@ -121,18 +120,36 @@ func runSetUser(cmd *cobra.Command, args []string) {
...
+	var hashed []byte
+	switch password {
+	case "":
+		hashed, err = security.PromptForPasswordAndHash()
+		if err != nil {
+			panic(err)
+		}
+	case "-":
+		scanner := bufio.NewScanner(os.Stdin)
+		scanner.Scan()
```
Comment (tamird) [inline cli/user.go]: do you need to inspect this return value?

**Comment by tamird** [comment]
Can we squash these commits down?

File: cli/user.go (hunk 2 above)
Comment (tschottdorf) [inline cli/user.go]: yep.

File: cli/flags.go (hunk 1 above)
Comment (nak3) [inline cli/flags.go]: It's much better. I replaced it with yours.

File: cli/user.go (hunk 3 above)
Comment (nak3) [inline cli/user.go]: I think we don't need to inspect the return value. However, I noticed that it allows empty password. So I added empty password check below.

Comment (tschottdorf) [inline cli/user.go]: I think @tamird has a point though. If you enter `<mypassword><EOF>` then that should be ok, but if you enter `<mypassword><ENTER>someotherstuff<EOF>` then maybe that should return an error. So you'd want `Scan()` to return `false` or error out with a message such as `extra bytes read after newline`. The case in which that could make a difference is if someone is auto-generating passwords and doesn't know newlines aren't permitted (ok, quite exotic, but still).

**Comment by tschottdorf** [comment]
I just tested your code locally, and I think you still need to fix and a rebase (`make check` isn't clean due to missing comment on `HashPassword`).

File: cli/user.go (hunk 3 above)
Comment (nak3) [inline cli/user.go]: OK, I see. I updated it.

**Comment by nak3** [comment]
@tschottdorf Sorry it is my mistake. I fixed and passed `make check`.

File: cli/user.go [hunk 4]
```diff
@@ -121,18 +121,48 @@ func runSetUser(cmd *cobra.Command, args []string) {
...
+	var hashed []byte
+	switch password {
+	case "":
+		hashed, err = security.PromptForPasswordAndHash()
+		if err != nil {
+			panic(err)
+		}
+	case "-":
+		scanner := bufio.NewScanner(os.Stdin)
+		if scanner.Scan() == true {
```
Comment (tamird) [inline cli/user.go]: `== true` is a silly pattern, `scanner.Text()` allocates, and there's no need to `util.Erorrf` before calling `panic()`. How about:
```go
	scanner := bufio.NewScanner(os.Stdin)
	if scanner.Scan() {
		if b := scanner.Bytes(); len(b) > 0 {
			hashed, err = security.HashPassword(b)
			if err != nil {
				panic(err)
			}
			if scanner.Scan() {
				panic("multiline passwords are not permitted")
			}
			if err := scanner.Err(); err != nil {
				panic(err)
			}

			break // Success.
		}
	} else {
		if err := scanner.Err(); err != nil {
			panic(err)
		}
	}

	panic("empty passwords are not permitted")
```

Comment (nak3) [inline cli/user.go]: That's great.. Thank you. I updated to it.

**Comment by tschottdorf** [comment]
LGTM. Thanks @nak3!

---

Reviewed 3 of 3 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-:-K7UvKmoRWqpkwB79FIO:-1981340430)*

**Comment by tschottdorf** [comment]
You still haven't rebased for the last three days. Please do that before we merge.

**Comment by nak3** [comment]
Oh, sorry! I have done.

**Comment by nak3** [comment]
I think it is successfully merged. But please let me know, if I misunderstood.

**Comment by tschottdorf** [comment]
LGTM. Thanks for contributing!

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3603#-:-K7_gNsVY5GYHBDw3jvj:1673116436)*

**Pull request merged**